src/FastSerialization/StreamReaderWriter.cs

```diff
@@ -693,38 +702,68 @@ protected override void Dispose(bool disposing)
                 positionInStream += (uint)position;
                 endPosition = 0;
                 position = 0;
-                // if you are within one read of the end of file, go backward to read the whole block.  
-                uint lastBlock = (uint)(((int)inputStream.Length - bytes.Length + align) & ~(align - 1));
-                if (positionInStream >= lastBlock)
-                {
-                    position = (int)(positionInStream - lastBlock);
-                }
-                else
+                if (inputStream.CanSeek)
                 {
-                    position = (int)positionInStream & (align - 1);
-                }
+                    // if you are within one read of the end of file, go backward to read the whole block.  
+                    uint lastBlock = (uint)(((int)inputStream.Length - bytes.Length + align) & ~(align - 1));
+                    if (positionInStream >= lastBlock)
+                    {
+                        position = (int)(positionInStream - lastBlock);
+                    }
+                    else
+                    {
+                        position = (int)positionInStream & (align - 1);
+                    }
 
-                positionInStream -= (uint)position;
+                    positionInStream -= (uint)position;
+                }
             }
 
             Debug.Assert(positionInStream % align == 0);
             lock (inputStream)
             {
-                inputStream.Seek(positionInStream + endPosition, SeekOrigin.Begin);
-                for (; ; )
+                // We need to get the stream positioned at (positionInStream + endPosition)
+                // Seekable streams: Easy we can seek
+                // Non-seekable streams: We need to read forward. We already did error checking
+                //                       in Goto() to ensure that the stream movement is going
+                //                       forward, not backwards.
+                if(inputStream.CanSeek)
+                {
+                    inputStream.Seek(positionInStream + endPosition, SeekOrigin.Begin);
+                }
+                else
+                {
+                    uint seekForwardDistance = (uint)((positionInStream + endPosition) - inputStreamBytesRead);
```

Comment:
```suggestion
                    // For non-seekable streams we can only move forward.   A check in the Goto routine
                    // insures that we don't try to move backwards.   Here is where we use it.  
                    Debug.Assert( (uint)(positionInStream + endPosition)) >= inputStreamBytesRead);
                    uint seekForwardDistance = (uint)((positionInStream + endPosition) - inputStreamBytesRead);
```

Reply:
Another worry with streaming is that you can definitely imagine it exceeding the 4GB stream size.   We should change things so that variables like inputStreamBytesRead and positionInStream are longs and to avoid issues with overflow.

Reply:
Hmm, it is worse than that.   The upper level event-pipe logic uses StreamLabels (which are 32 bit), and thus would have issues if the stream gets larger than 4GB.     Fixing it at the StreamReader level is not sufficient.   

I was tempted when I first saw the logic to simply disallow Goto at all (thus very little change is needed), but saw that we DO use Goto (but we don't need to).    We do have a extension method called Skip() that can avoid these problems.    Thus my recommendation is

1. Change EventPipeEventSource so that it uses Skip instead of GoTo.
2. Promote Skip to be a instance method on a StreamReader.
3. Make Goto uniformly fail (throw exception) on non-seekable streams.
4. Make Skip work properly regardless of stream size (basically make some of these variables long instead of uint).

Reply:
> Hmm, it is worse than that...

Yeah, I believe there are a variety of problems that need to be resolved before this can scale up. I wasn't planning to tackle those issues just yet as I only needed something good enough to preview dotnet-counters which has small data sizes. In addition to the issues you spotted I noticed we've also got [caching](https://github.com/Microsoft/perfview/blob/master/src/FastSerialization/FastSerialization.cs#L1887). 

> saw that we DO use Goto (but we don't need to)

Goto() isn't always forward, it appears to depend on the data we are reading and the mode the deserializer is operating in. For example:
- [Peeking](https://github.com/Microsoft/perfview/blob/765f5908c46b55366cc6420db3c84e1879c4d43a/src/FastSerialization/StreamReaderWriter.cs#L651)
- [Recovering after invalid data](https://github.com/Microsoft/perfview/blob/master/src/FastSerialization/FastSerialization.cs#L1076)
 - [Returning after reading a forward reference](https://github.com/Microsoft/perfview/blob/master/src/FastSerialization/FastSerialization.cs#L1318)
 - [Reading a back-reference](https://github.com/Microsoft/perfview/blob/master/src/FastSerialization/FastSerialization.cs#L1811)

The PR avoids the first two issues by allowing backwards movement [within the range of the Fill buffer](https://github.com/Microsoft/perfview/blob/765f5908c46b55366cc6420db3c84e1879c4d43a/src/FastSerialization/StreamReaderWriter.cs#L651). It relies on the netperf format not to use forward or backward references to handle the other cases.

> 3. Make Goto uniformly fail (throw exception) on non-seekable streams.

I'm not sure what value this would get us? It will require lots of places that are calling Goto to be refactored and the places that are legitimately back-tracking inside the fill buffer for [peeking and error recovery](https://github.com/Microsoft/perfview/blob/master/src/FastSerialization/FastSerialization.cs#L1630) will now need algorithmic changes or a 2nd level of buffering.

The other ones seem like fine improvements, though I propose doing them in a future PR for expedience in the current work.